hal/spi_api.h

```diff
@@ -19,220 +18,345 @@
-    struct buffer_s tx_buff; /**< Tx buffer */
-    struct buffer_s rx_buff; /**< Rx buffer */
-} spi_t;
+    PinName ss; /**< Slave select pin. */
+    PinName miso; /**< Master In Slave our or D1 in DUAL and QUADIO mode */
+    /** Master Out Slave In, D2 in DUAL and QUARD-IO mode or may not be connected in 3-wires mode.*/
+    PinName mosi;
+    PinName mclk; /**< Master Clock pin. */
+    PinName d3; /**< Used in QUADIO mode. */
+    PinName d4; /**< Used in QUADIO mode. */
+} spi_pins_t;
 
-#else
-/** Non-asynch SPI HAL structure
+/**
+ * This structure groups all initialisation parameters required by a SPI interface.
  */
-typedef struct spi_s spi_t;
+typedef struct {
+    spi_pins_t pins; /**< Pins to use for this channel. */
 
-#endif
+    /**< This symbol will be sent as a place holder in full-duplex mode in case more symbols are to
+         be read than to be sent. */
+    uint32_t fill_symbol;
+    uint32_t clock_frequency; /**< MCLK frequency in Hz. */
 
-#ifdef __cplusplus
-extern "C" {
-#endif
+    uint32_t word_length; /**< Length of a symbol in bit. */
+    bool is_master; /**< True to configure the device in Master mode. */
+    bool msb_first; /**< True to send/receive the most significant bit first. */
+    bool clock_phase; /**< True if data line is valid when leaving active state. */
+    bool clock_polarity; /**< True if the clock's rest state is high (+Vcc). */
+    bool continuous_mode; /**< True to use the continuous mode. */
+
+    spi_mode_t  mode; /**< Transmission mode. See spi_mode_t. */
+} spi_init_t;
 
 /**
- * \defgroup hal_GeneralSPI SPI Configuration Functions
- * @{
+ * This enumerates the possible result of the spi_init function.
  */
+typedef enum {
+    SPI_RESULT_OK, /**< Operation successful. */
+    SPI_RESULT_CONFIG_UNSUPPORTED, /**< The required parameters are not supported by the device. */
+    SPI_RESULT_INVALID_PARAM, /**< The given parameter(s) is/are invalid. */
+    SPI_RESULT_DEVICE_BUSY, /**< The requested peripheral/SS p)in is already initialised. */
+} spi_result_t;
 
-/** Initialize the SPI peripheral
+/**
+ * This is describes the capabilities of a SPI channel.
+ * The elements of this structure are independent. This means that a supported feature is supported
+ * regardless of other feature settings.
  *
- * Configures the pins used by SPI, sets a default format and frequency, and enables the peripheral
- * @param[out] obj  The SPI object to initialize
- * @param[in]  mosi The pin to use for MOSI
- * @param[in]  miso The pin to use for MISO
- * @param[in]  sclk The pin to use for SCLK
- * @param[in]  ssel The pin to use for SSEL
+ * @note minimum_frequency and maximum_frequency are used by tests to assess that at least these
+ * frequencies are supported. The device may be able to achieve lower and/or faster frequency.
+ * @warning Only minimum_frequency and maximum_frequency are assessed during tests. Your device may
+ * not be able to generate the whole range between these limits.
```

Comment:
Please explain which tests and when is frequency range derived? If we cannot share correct capabilities of device then what is the point of this test?

Reply:
as discussed with @c1728p9 this will change to make this a strong requirement. All frequencies in the range have to be supported (with measure deviation).

Reply:
Won’t it be convenient to get the range from target device, at least in case of master. I understand testing in case of slave devices since users can use any slave device. But if target device is master, it should supply supported frequency range.